### DISCUSSION TITLE
Any interest in better date handling

### DISCUSSION THREAD

**Discussion by willsmithorg** [autogluon/autogluon#1433] [Ideas]
Currently dates - when recognised as such - are just converted to datetimes in milliseconds (I believe).

Is there any interest in separating them into components, e.g. year, month, day, hour, minute, second, ...?
I work in the finance industry and there are some events that happen at the end of each quarter of a year, for example.  Having month separated out would make that easier to spot.

If this sounds useful I could work on it myself and submit a PR.

Also - more advanced - strings in finance often have embedded dates and numbers.  For example, this is the description of a bond: "FED.HOME LOAN MGE. 2000 6 3/4% 15/03/31 S"    Fortunately the coupon rate (6.75%) and maturity date (15-Mar-2031) are in additional columns in my dataset.  But in general, text fields in tabular data often encode several fields.  Here's another one, stocks and bonds are often identified by an "ISIN", examples are US171232AE11 and CA683234QT85.  The first 2 characters tells you the country of issuance, which can then help you predict other features.  Do we expect users to do this level of feature engineering themselves, to create an additional feature with the first 2 characters only - or is the ambition of Autogluon to increasingly have domain-aware generators that have domain knowledge and recognise features as relevant to that domain?  If we want increasingly clever feature engineering, maybe we should detect that a text column, for example, can be split into substrings where the position is important - unlike the existing n-gram handling where the substring can be anywhere.  Another example - airline flight codes are in the format \w{2}\d{1,4}.  Separating out the \w{2} would identify the airline.

**Comment by Innixma** [comment]
Thanks for your comments! I agree that datetime handling can be done better by extracting additional information from the dates. I've opened #1436 to track this enhancement, and feel free to respond to the issue if you are interested in contributing the functionality! I'm happy to help if you have any questions along the way.

Regarding the more advanced scenarios: The tricky parts when trying to extract complex information from free-form text are:

1. Processing Time: Specialized text extraction can be expensive. If the feature only exists in very few datasets, then it will slow down every dataset that doesn't need that functionality.
2. False Positives: The more special functionality we add that checks strings for matches the more likely it will find it incorrectly or the feature won't be helpful on a particular dataset. 


So for these specialized techniques, the way we can get them into AutoGluon is:
1. Be fast or widely applicable.
2. Unlikely to generate false positives.

OR

1. Require user to explicitly specify the generator (not enabled by default).

Happy to continue discussion on this point as it is very interesting and the long-term vision of AutoML should be to handle these complex cases.

**Reply by willsmithorg** [reply]
Hi Innixma.

Firsty, thanks for reaching out.  I see you checking in huge amounts of code (including on weekends) and still finding time to engage in detail with external contributors and users.  Impressive!


I went through the 39 autoML classification datasets one by one, and what I found rather surprised me:
- around 30 have only floating point numeric features
- only about 6 have any strings whatsoever, and these are mostly trivially turned into categoricals
- almost all datasets have been 'pre-feature-engineered' e.g. one-hot encoding etc, strings to numerical categoricals

This seems to me the opposite of where we'd expect a "citizen data scientist" to be - more likely they will have a tabular dataset straight out of some existing business logic, with a mix of integers, floating points, categoricals as integers,  short strings (some simpler categoricals, some embedding more complex information that requires domain knowledge to fully understand) and longer text.   At best, we can ask them whether their problem is a classification or a regression problem - maybe even that they are not sure about if their target 'label' column is numeric.   I  believe we should expect them to do no feature engineering at all.

So I guess it comes down to hand-estimated heuristics initially, like you used in the date transformer class - do x% of the values look like dates?  Then it's a date.  Do most of the strings in this column have the same separator, like a dash or a dot, or a predictable junction between alphabetics and numerics?  If so, let's split on that separator/junction and add additional features.

One area autogluon already has a heuristic that may be going wrong is the numeric vs categorical threshold.  It hit me on a dataset - we have countries coded as integers and there are ~100.   I wanted that categorical, but it defaulted to numeric.  Perhaps there should be two thresholds (definitely categorical, definitely numeric) , and those that fall in the middle, we duplicate the feature into the dataset as both categorical and numeric features.  Some models (e.g. LightGBM) need to be told what is categorical.  I got a good increase in my model accuracy when I hand-crafted my country  as categorical - by coercing with d['_country'] = 'x' + d['country'].astype(str).  A less trained user wouldn't know to try this.

Maybe as Amazon employees you guys have insight into what people are submitting to  Sagemaker for tabular problems? Maybe some summary stats would be good for analysis - what % of problems contain text features, what is the distribution of the length of the text features.

What I propose is that I actively hunt out datasets that have non-trivial text features (high cardinality but not many spaces, so not freeform text), and I try to setup the benchmarking code to include these datasets, then we run competitions with more vs less feature engineering on text features.  So far I tried to set up the benchmarking and failed (too many shell scripts to run on various machines both local and EC2, I kept getting errors).  Is there any way you guys could set up a CI environment to run the benchmarks on demand and report results?  That would also protect against accuracy regressions between versions.

Thanks again for your time.

**Reply by Innixma** [reply]
Hi Will,

You bring up a lot of great points here!

Regarding the 39 AutoML datasets, you are correct that they are fairly 'tame'. This is because historically, AutoGluon (and to an extent H2O) was the first AutoML framework to support raw data input. TPOT, AutoWeka, Auto-Sklearn, etc. did not support this as they were primarily focused on the modeling side of AutoML rather than the feature cleaning / data prep side. As AutoMLBenchmark pre-dates AutoGluon, the benchmark (and OpenML's datasets) focus on these clean datasets. Further, publicly available messy datasets are quite rare (although we hope to change that in the coming years).

Your description of 'citizen data scientist' is exactly right. This is how AutoGluon is used in the grand majority of cases within Amazon, and likely elsewhere. This is a large reason why I architected AutoGluon to handle raw data, as most users don't know how to properly clean the data themselves (or they would rather something did it for them).

Inferring features via heuristics is one approach, and is what I have done so far. Another approach is to use meta-learning to have models which predict the feature types (such as in [this paper](https://adalabucsd.github.io/papers/TR_2020_SortingHat.pdf) which compares against Pandas, AutoGluon, etc. for the purposes of feature type inference). I haven't personally dived very deeply into this topic, but it could be promising. Regarding extracting structured information from text, this is definitely do-able but relies on many assumptions on the structure of the text, and therefore could get complicated fast. We currently have SOTA results by leveraging transformer architectures to parse multi-modal data implemented in AutoGluon ([Paper](https://arxiv.org/pdf/2111.02705.pdf)). This approach side-steps the entire issue by putting the burden of parsing / understanding the text to the transformer. Of course, transformers have various limitations and won't understand structured data as well as a tabular model, so its not a full solution.

Re Int vs Categorical, I agree. I've been aware that incorrect identification of a categorical encoded as int to be an int is common in AutoGluon for a long time, but I haven't set aside time dedicated to fixing it. I think there are some pretty simple heuristic rules to apply that would improve detection, such as `if values are from 0-N, without any missing values between 0 and N, and N < 100, then it is a categorical`.

Re setting up the benchmarking: If you are trying to use the benchmarking code from autogluon-benchmarking, I apologize. This code is very old and likely will not work without a great deal of effort. We are currently [working](https://github.com/openml/automlbenchmark/issues/434) on [mechanisms](https://github.com/openml/automlbenchmark/issues/231) to make benchmarking easily available to contributors (and ourselves). At the moment, it is [fairly indepth](https://github.com/Innixma/autogluon-benchmark/blob/master/examples/automlbenchmark/README_automlbenchmark.md), and I would not recommend trying to follow the steps until it is in a more friendly state. Currently I manually run benchmarks for most PRs via a series of scripts I've written, but I want to change this to be linked to GitHub Actions so we can simply trigger the benchmark run for a given PR on-demand.

If you'd like to identify datasets that would be useful to benchmark for AutoML, that would be fantastic!

As a final note, I am going on vacation starting tomorrow and likely won't be responding to issues with the same frequency until I return in early Jan. Happy Holidays!
